sql/src/main/java/io/crate/execution/dml/upsert/ReturnValueGen.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * Licensed to Crate under one or more contributor license agreements.
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.  Crate licenses this file
+ * to you under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.execution.dml.upsert;
+
+import io.crate.data.Input;
+import io.crate.execution.engine.collect.CollectExpression;
+import io.crate.expression.InputFactory;
+import io.crate.expression.reference.Doc;
+import io.crate.expression.reference.DocRefResolver;
+import io.crate.expression.symbol.Symbol;
+import io.crate.metadata.Functions;
+import io.crate.metadata.TransactionContext;
+import io.crate.metadata.doc.DocTableInfo;
+
+import javax.annotation.Nullable;
+import java.util.List;
+
+import static io.crate.common.collections.Lists2.map;
+
+final class ReturnValueGen {
+
+    private final InputFactory.Context<CollectExpression<Doc, ?>> cntx;
+
+    ReturnValueGen(Functions functions, TransactionContext txnCtx, DocTableInfo table, Symbol[] returnValues) {
+        InputFactory.Context<CollectExpression<Doc, ?>> cntx = new InputFactory(functions).ctxForRefs(
+            txnCtx, new DocRefResolver(table.partitionedBy()));
+        cntx.add(List.of(returnValues));
+        this.cntx = cntx;
+    }
+
+    @Nullable
+    Object[] generateReturnValues(Doc doc) {
+        cntx.expressions().forEach(x -> x.setNextRow(doc));
+        List<Object> result = map(cntx.topLevelInputs(), Input::value);
+
+        return result.isEmpty() ? null : result.toArray();
```

Comment:
How about something like:

```
if (ctx.topLevelInputs.isEmpty()) {
    return null;
}
Object[] result = new Object[ctx.topLevelInputs().size()];
for (int i = 0; i < result.length; i++) {
    result[i] = ctx.topLevelInputs().get(i).value();
}
return result;
```

Would avoid some indirections.

Maybe also pre-assign the topLevelInputs + expressions to an attribute in the CTOR and don't keep a reference to the context itself. Allows to GC `InputFactory.Context` earlier which pulls in some other objects.

Reply:
👍